qualtran/bloqs/factoring/mod_arithmetic.py

```diff
@@ -0,0 +1,224 @@
+
+    def bloq_counts(self, ssa: Optional['SympySymbolAllocator'] = None) -> Set['BloqCountT']:
+        return {(2 * self.bitsize, Toffoli())}
+
+    def build_composite_bloq(
+        self, bb: 'BloqBuilder', x: SoquetT, p: SoquetT
+    ) -> Dict[str, 'SoquetT']:
+        return NotImplemented
+
+    def short_name(self) -> str:
+        return f'x = -x mod p'
+
+
+@frozen
+class ModDbl(Bloq):
+    r"""An n-bit modular negation gate.
+
+    Implements $U|x\rangle|p\rangle \rightarrow |2 * x mod p\rangle|p\rangle$ using $2n T$ gates.
+
+    Args:
+        bitsize: Number of bits used to represent each integer.
+
+    Registers:
+        x: A bitsize-sized input register (register x above).
+        p: A bitsize-sized input register (register p above).
+
+    References:
+        [How to compute a 256-bit elliptic curve private key with only 50 million Toffoli gates](https://arxiv.org/abs/2306.08585)
+    """
+
+    bitsize: Union[int, sympy.Expr]
+
+    @cached_property
+    def signature(self) -> 'Signature':
+        return Signature([Register('x', bitsize=self.bitsize), Register('p', bitsize=self.bitsize)])
+
+    def bloq_counts(self, ssa: Optional['SympySymbolAllocator'] = None) -> Set['BloqCountT']:
+        return {(2 * self.bitsize, Toffoli())}
+
+    def build_composite_bloq(
+        self, bb: 'BloqBuilder', x: SoquetT, p: SoquetT
+    ) -> Dict[str, 'SoquetT']:
+        return NotImplemented
+
+    def short_name(self) -> str:
+        return f'x = 2 * x mod p'
+
+
+@frozen
+class ModMult(Bloq):
+    r"""An n-bit modular multiplication gate.
+
+    Implements $U|x\rangle|y\rangle|p\rangle|0\rangle|0\rangle \rightarrow |x\rangle|y\rangle|p\rangle|garbage\rangle|x * y mod p\rangle$ using $2.25n^2 + 9n T$ gates.
+
+    Args:
+        bitsize: Number of bits used to represent each integer.
+
+    Registers:
+        x: A bitsize-sized input register (register x above).
+        y: A bitsize-sized input register (register y above).
+        p: A bitsize-sized input register (register p above).
+        garbage: A bitsize-sized input register (register garbage above).
+        out: A bitsize-sized input register holding the output of the modular multiplication.
+
+    References:
+        [How to compute a 256-bit elliptic curve private key with only 50 million Toffoli gates](https://arxiv.org/abs/2306.08585)
+    """
+
+    bitsize: Union[int, sympy.Expr]
+
+    @cached_property
+    def signature(self) -> 'Signature':
+        return Signature(
+            [
+                Register('x', bitsize=self.bitsize),
+                Register('y', bitsize=self.bitsize),
+                Register('p', bitsize=self.bitsize),
+                Register('garbage', bitsize=self.bitsize),
+                Register('out', bitsize=self.bitsize),
+            ]
+        )
+
+    def bloq_counts(self, ssa: Optional['SympySymbolAllocator'] = None) -> Set['BloqCountT']:
+        return {(2.25 * (self.bitsize**2) + 9 * self.bitsize, Toffoli())}
```

Comment:
This must be the ceiling of this value. I imagine they fit a line to get the slope, but the counts of bloq counts should integers.

Reply:
Yes, I believe this is true. I can truncate the 2.25 for now to ensure it is always an integer. I hope to have a more exact number when I implement the bloq decomposition.